mysqltest/docker.go

```diff
@@ -0,0 +1,193 @@
+)
+
+var nopCloser = io.NopCloser(nil)
+
+// start starts a docker container running a MySQL server. A struct is returned describing how to
+// connect to MySQL, along with a cleanup function that should be called once all tests have
+// finished.
+//
+// The returned Closer should be called in every case, even if this function returns an error. This
+// ensures that the Docker container will be cleaned up if the error occurred after the container
+// was created. The Closer will never be nil.
+//
+// Since the startup time for this MySQL container is about 20 seconds, we share the container among
+// every test. Each test should use a randomly-created database/schema name to avoid collisions
+// between tests.
+//
+// This function installs a signal handler for SIGTERM, SIGKILL, and SIGINT that attempts to clean
+// up the Docker container, then runs os.Exit(1). Since the signal handler kills the process, any
+// other custom signal handlers that are installed may not get a chance to run.
+//
+// Docker must be installed on localhost for this to work. No environment vars are needed.
+func start(conf *config) (ConnInfo, io.Closer, error) {
+	port, closer, err := startContainer(conf)
+	if err != nil {
+		return ConnInfo{}, nopCloser, err
+	}
+
+	return ConnInfo{
+		Username: "root",
+		Password: password,
+		Hostname: "localhost",
+		Port:     port,
+	}, closer, nil
+}
+
+// Runs a MySQL docker container and returns its TCP port, along with a cleanup function that stops
+// the container.
+func startContainer(conf *config) (string, io.Closer, error) {
+	pool, err := dockertest.NewPool("")
+	if err != nil {
+		return "", nopCloser, fmt.Errorf("dockertest.NewPool(): %w", err)
+	}
+	container, err := runContainer(conf, pool)
+	if err != nil {
+		return "", nopCloser, err
+	}
+	closer := newContainerCloser(pool, container)
+	if err := container.Expire(uint(conf.killAfterSec)); err != nil {
+		return "", closer, fmt.Errorf("resource.Expire(): %w", err)
+	}
+	outPort, err := waitUntilUp(conf, pool, container)
+	if err != nil {
+		return "", closer, err
+	}
+	return outPort, closer, nil
+}
+
+// Starts the mysql container and returns a Resource that points to it.
+func runContainer(conf *config, pool *dockertest.Pool) (*dockertest.Resource, error) {
+	// pulls an image, creates a container based on it and runs it
+	container, err := pool.RunWithOptions(&dockertest.RunOptions{
+		Repository: "mysql",
+		Tag:        conf.mySQLVersion,
+		Env:        []string{"MYSQL_ROOT_PASSWORD=" + password},
+	}, func(config *docker.HostConfig) {
+		config.AutoRemove = true // remove storage after container exits
+		config.RestartPolicy = docker.RestartPolicy{Name: "no"}
+	})
+	if err != nil {
+		var extraMsg string
+		switch {
+		case strings.Contains(err.Error(), "no such file"):
+			extraMsg = `. Please install docker: 
+		Instructions for gLinux: https://g3doc.corp.google.com/cloud/containers/g3doc/glinux-docker/install.md
```

Comment:
@sethvargo @crwilcox I'm not sure if it's ok to have link to g3doc in this public repo

Reply:
Aha, good catch, will fix shortly.